src/app/features/worklog/worklog-export/worklog-export.util.ts

```diff
@@ -4,190 +4,188 @@ import { msToString } from '../../../ui/duration/ms-to-string.pipe';
+ */
+const handleTaskGroup = (data: WorklogExportData, groupBy: WorklogGrouping): ItemsByKey<RowItem> => {
+  const taskGroups: ItemsByKey<RowItem> = {};
+  for (const task of data.tasks) {
+    if (skipTask(task, groupBy)) { continue; };
+    const taskFields = getTaskFields(task, data);
+    const dates = sortDateStrings(Object.keys(task.timeSpentOnDay));
+    taskGroups[task.id] = {
+      dates,
+      timeEstimate: task.timeEstimate,
+      timeSpent: Object.values(task.timeSpentOnDay).reduce((acc, curr) => acc + curr, 0),
+      workStart: 0,
+      workEnd: 0,
+      ...taskFields
+    };
+  }
+  return taskGroups;
+};
 
-    group.notes = group.notes.map((note) => {
-      return note.replace(/\n/g, ' - ');
+/**
+ * For each task creates a new rowItem without needing to push to previous taskGroups, unlike handleDateGroup.
+ * We're still creating a map since we will use the key for sorting in the next step.
+ */
+const handleWorklogGroup = (data: WorklogExportData): ItemsByKey<RowItem> => {
+  const taskGroups: ItemsByKey<RowItem> = {};
+  for (const task of data.tasks) {
+    Object.keys(task.timeSpentOnDay).forEach(day => {
+      const groupKey = task.id + '_' + day;
+      const taskFields = getTaskFields(task, data);
+      taskGroups[groupKey] = {
+        dates: [day],
+        timeEstimate: task.subTaskIds.length > 0 ? 0 : task.timeEstimate,
+        timeSpent: task.subTaskIds.length > 0 ? 0 : task.timeSpentOnDay[day],
+        workStart: data.workTimes.start[day],
+        workEnd: data.workTimes.end[day],
+        ...taskFields
+      };
     });
+  }
+  return taskGroups;
+};
 
-    group.projects = unique(group.projects);
-    group.projects = group.projects.map((pId: string) => {
-      return (allProjects.find(project => project.id === pId) as ProjectCopy).title;
-    });
+/**
+ * Unfolds task into taskFields while mapping id's to titles, and minor formatting
+ */
+const getTaskFields = (task: WorklogTask, data: WorklogExportData): TaskFields => {
+  const titlesWithSub = [task.title];
+  const titles = task.parentId ?
+    [(data.tasks.find(ptIN => ptIN.id === task.parentId) as WorklogTask).title] :
+    [task.title];
 
-    group.tags = unique(group.tags);
-    group.tags = group.tags.map((tId: string) => {
-      return (allTags.find(tag => tag.id === tId) as TagCopy).title;
-    });
+  const notes = task.notes ? [task.notes.replace(/\n/g, ' - ')] : [];
+  const projects = task.projectId ?
+    [(data.projects.find(project => project.id === task.projectId) as ProjectCopy).title] :
+    [];
 
-    rows.push(group);
-  }));
+  // by design subtasks don't have tags, so we must set its parent's tags
+  let tags = task.parentId !== null ?
+    (data.tasks.find(t => t.id === task.parentId) as WorklogTask).tagIds :
+    task.tagIds;
+  tags = tags.map( tagId => (data.tags.find(tag => tag.id === tagId) as TagCopy).title);
 
-  return rows;
+  const tasks = [ task ];
```

Comment:
This, now, is only needed because of what happens in line ~209. I'm not sure about the intention of that part of code, especially the part of `row.tasks[0]`... In my opinion this should be handled inside handleWorklogGroup().

Reply:
I agree with you, but I can also live with the current approach.